File: java/say/src/main/java/Say.java
public class Say {

    public String say(long number) {

        if (number < 0 || number > 999_999_999_999L) {
            throw new IllegalArgumentException();
        }

        if (number == 0) {
            return "zero";
        }

        int b = (int) (number / 1_000_000_000);
        int m = (int) (number % 1_000_000_000) / 1_000_000;
        int t = (int) (number % 1_000_000);
        String res = billions(b);
        res += m > 0 ? " " + millions(m) : "";
        res += t > 0 ? " " + thousands(t) : "";

        return res.trim();
    }

    private String billions(int number) {
        return number > 0 ? hundreds(number) + " billion" : "";
    }

    private String millions(int number) {
        return number > 0 ? hundreds(number) + " million" : "";
    }

    private String thousands(int number) {
        int tn = number / 1000;
        return (tn > 0 ? hundreds(tn) + " thousand" : "")
                + ((tn > 0 && number % 1000 > 0) ? " " : "")
                + ((number % 1000 > 0) ? hundreds(number % 1000) : "");
    }

    private String hundreds(int number) {
        int n = number / 100;
        String t = number % 100 != 0 ? " " + tens(number % 100) : "";
        return n == 0 ? tens(number) : units(n) + " hundred" + t;
    }

    private String tens(int number) {
        int n = number / 10;
        String u = number % 10 != 0 ? "-" + units(number % 10) : "";
        return switch (n) {
            case 1 -> teens(number);
            case 2 -> "twenty" + u;
            case 3 -> "thirty" + u;
            case 4 -> "forty" + u;
            case 5 -> "fifty" + u;
            case 6 -> "sixty" + u;
            case 7 -> "seventy" + u;
            case 8 -> "eighty" + u;
            case 9 -> "ninety" + u;
            default -> units(number);
        };
    }

    private String teens(int number) {
        return switch (number) {
            case 10 -> "ten";
            case 11 -> "eleven";
            case 12 -> "twelve";
            case 13 -> "thirteen";
            case 14 -> "fourteen";
            case 15 -> "fifteen";
            case 16 -> "sixteen";
            case 17 -> "seventeen";
            case 18 -> "eighteen";
            case 19 -> "nineteen";
            default -> "";
        };
    }

    private String units(int number) {
        return switch (number) {
            case 1 -> "one";
            case 2 -> "two";
            case 3 -> "three";
            case 4 -> "four";
            case 5 -> "five";
            case 6 -> "six";
            case 7 -> "seven";
            case 8 -> "eight";
            case 9 -> "nine";
            default -> "";
        };
    }
}

File: javascript/say/say.js
export const say = (number) => {

    if (number < 0 || number > 999_999_999_999) {
        throw new Error('Number must be between 0 and 999,999,999,999.');
    }

    let s = '';
    if (number === 0) {
        return 'zero';
    }

    s = oneToOneThousand(number);

    let th = number % 1_000_000;
    th = Math.trunc(th / 1000);
    if (th > 0) {
        s = oneToOneThousand(th) + ' thousand ' + s;
    }

    let m = number % 1_000_000_000;
    m = Math.trunc(m / 1_000_000);
    if (m > 0) {
        s = oneToOneThousand(m) + ' million ' + s;
    }

    let b = number % 1_000_000_000_000;
    b = Math.trunc(b / 1_000_000_000);
    if (b > 0) {
        s = oneToOneThousand(b) + ' billion ' + s;
    }

    return s.trim();

};

const oneToOneThousand = (number) => {        

    let s = '';

    let u = number % 10;
    s += units(u);

    let t = number % 100;

    if (t > 10 && t < 20) {
        s = teens(t);
    } else if (t >= 20) {
        t -= u;
        if (u == 0) {
            s = tens(t);
        } else {
            s = tens(t) + '-' + s;
        }
    }

    let h = number % 1000;
    h = Math.trunc(h / 100);
    if (h > 0) {
        s = units(h) + ' hundred ' + s;
    }

    return s.trim();
};

const units = (number) => {
    switch (number) {
        case 1:
            return 'one';
        case 2:
            return 'two';
        case 3:
            return 'three';
        case 4:
            return 'four';
        case 5:
            return 'five';
        case 6:
            return 'six';
        case 7:
            return 'seven';
        case 8:
            return 'eight';
        case 9:
            return 'nine';
        default:
            return '';

    }
};

const teens = (number) => {

    switch (number) {
        case 11:
            return 'eleven';
        case 12:
            return 'twelve';
        case 13:
            return 'thirteen';
        case 14:
            return 'fourteen';
        case 15:
            return 'fifteen';
        case 16:
            return 'sixteen';
        case 17:
            return 'seventeen';
        case 18:
            return 'eighteen';
        case 19:
            return 'nineteen';
        default:
            return '';
    }
};

const tens = (number) => {

    switch (number) {
        case 10:
            return 'ten';
        case 20:
            return 'twenty';
        case 30:
            return 'thirty';
        case 40:
            return 'forty';
        case 50:
            return 'fifty';
        case 60:
            return 'sixty';
        case 70:
            return 'seventy';
        case 80:
            return 'eighty';
        case 90:
            return 'ninety';
        default:
            return '';
    }
}
